clash_lib/src/proxy/tuic/types.rs

```diff
@@ -0,0 +1,274 @@
+                    (conn.await?, None)
+                };
+
+                Ok((conn, zero_rtt_accepted))
+            };
+
+            match connect_to.await {
+                Ok((conn, zero_rtt_accepted)) => {
+                    return Ok(TuicConnection::new(
+                        conn,
+                        zero_rtt_accepted,
+                        self.udp_relay_mode,
+                        self.uuid,
+                        self.password.clone(),
+                        self.heartbeat,
+                        self.gc_interval,
+                        self.gc_lifetime,
+                    ));
+                }
+                Err(err) => last_err = Some(err),
+            }
+        }
+        Err(last_err.unwrap_or(anyhow!("dns resolve")))
+    }
+}
+
+#[derive(Clone)]
+pub struct TuicConnection {
+    pub conn: QuinnConnection,
+    pub inner: InnerConnection<tuic_quinn::side::Client>,
+    pub uuid: Uuid,
+    pub password: Arc<[u8]>,
+    pub remote_uni_stream_cnt: Counter,
+    pub remote_bi_stream_cnt: Counter,
+    pub max_concurrent_uni_streams: Arc<AtomicU32>,
+    pub max_concurrent_bi_streams: Arc<AtomicU32>,
+    pub udp_relay_mode: UdpRelayMode,
+    pub udp_sessions: Arc<AsyncRwLock<HashMap<u16, UdpSession>>>,
+}
+
+pub struct UdpSession {
+    pub incoming: tokio::sync::mpsc::Sender<UdpPacket>,
+    pub local_addr: ClashSocksAddr,
+}
+
+impl TuicConnection {
+    pub fn is_closed(&self) -> bool {
+        self.conn.close_reason().is_some()
+    }
+    fn new(
+        conn: QuinnConnection,
+        zero_rtt_accepted: Option<ZeroRttAccepted>,
+        udp_relay_mode: UdpRelayMode,
+        uuid: Uuid,
+        password: Arc<[u8]>,
+        heartbeat: Duration,
+        gc_interval: Duration,
+        gc_lifetime: Duration,
+    ) -> Self {
+        let conn = Self {
+            conn: conn.clone(),
+            inner: InnerConnection::<tuic_quinn::side::Client>::new(conn),
+            uuid,
+            password,
+            udp_relay_mode,
+            remote_uni_stream_cnt: Counter::new(),
+            remote_bi_stream_cnt: Counter::new(),
+            // TODO: seems tuic dynamicly adjust the size of max concurrent streams, is it necessary to configure the stream size?
+            max_concurrent_uni_streams: Arc::new(AtomicU32::new(32)),
+            max_concurrent_bi_streams: Arc::new(AtomicU32::new(32)),
+            udp_sessions: Arc::new(AsyncRwLock::new(HashMap::new())),
+        };
+
+        tokio::spawn(
+            conn.clone()
+                .init(zero_rtt_accepted, heartbeat, gc_interval, gc_lifetime),
+        );
+
+        conn
+    }
+    async fn init(
+        self,
+        zero_rtt_accepted: Option<ZeroRttAccepted>,
+        heartbeat: Duration,
+        gc_interval: Duration,
+        gc_lifetime: Duration,
+    ) {
+        tracing::info!("connection established");
+
+        // TODO reduct spawn
+        tokio::spawn(self.clone().tuic_auth(zero_rtt_accepted));
```

Comment:
do you need to clone for every task? could you do something like this https://github.com/Watfaq/clash-rs/blob/master/clash_lib/src/proxy/wg/wireguard.rs#L134

Reply:
For `heartbeat` and `collect_garbage` I would use tokio::select later, the Futures in them have [Cancellation Safety](https://docs.rs/tokio/latest/tokio/macro.select.html#cancellation-safety). The `tuic_auth` seems not have cancellation safety, I will look into it.

Reply:
Arc wraps `Self` for now.